files-community/Files
# Code Quality: Introduce details layout columns generator phase 2

## Summary

This is the phase 2 of creating a generator of columns for DetailsLayout page.

We expect that this will significantly reduce the burden on developers implementing new columns to actually make additional changes, significantly reduce the amount of XAML code in DetailsLayoutPage, keeping the codebase clean, and allowing us to easily refactor.

## Task Checklist

- Delete `ColumnsViewModel`
- Rename `FolderSettings` properties with `LayoutPreferencesManager`
- Added a new factory for the generator of columns
- Added a new setting named `DefaultColumnItems` collection in the app json settings
- Added a new converter named `DoubleToGridLength` to avoid relying on UI class in the model/item class.
- Added ViewModel for DetailsLayoutPage
- Moved columns-related methods/commands to the ViewModel

Note that there's a small amount of refactoring on the codebase.

- [x] Sort indicator behavior
- [x] Resizing behavior
- [ ] Show/Hide column toggle behavior
- [ ] Git repository folder
- [ ] Recycle Bin folder

## Steps To Validate Changes

1. Launch the app
2. Go to Desktop
3. Resize random column
4. Double click a resizer
5. Click a column to sort
6. Hide a column from the menu flyout
7. Go to Downloads
8. Go back to Desktop
9. See all settings you set is applied correctly
10. Go to a random folder with DetailsLayout
11. Try 3-9 randomly

## Known Issues

_N/A_

## PR Checklist

- [x] **Development**: #12505

## Screenshots

_N/A_

[^1]: If the request hasn't been agreed by the team, this work might be rejected. Make sure that you get approval from the team before opening any PR related the request.
[^2]: If you removed any en-US string resources, make sure that there are no references of those resources. When you add a new en-US string resources, do not make any changes on other languages' resources; [Crowdin](https://crowdin.com/project/files-app) will do that, instead.
